src/Magnum/DartIntegration/ConvertShapeNode.h

```diff
@@ -0,0 +1,82 @@
+#ifndef Magnum_DARTIntegration_ConvertShapeNode_h
+#define Magnum_DARTIntegration_ConvertShapeNode_h
+/*
+    This file is part of Magnum.
+
+    Copyright © 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018
+              Vladimír Vondruš <mosra@centrum.cz>
+    Copyright © 2018 Konstantinos Chatzilygeroudis <costashatz@gmail.com>
+
+    Permission is hereby granted, free of charge, to any person obtaining a
+    copy of this software and associated documentation files (the "Software"),
+    to deal in the Software without restriction, including without limitation
+    the rights to use, copy, modify, merge, publish, distribute, sublicense,
+    and/or sell copies of the Software, and to permit persons to whom the
+    Software is furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included
+    in all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+    DEALINGS IN THE SOFTWARE.
+*/
+
+/** @file
+ * @brief Function @ref Magnum::DartIntegration::convertShapeNode()
+ * @brief Class @ref Magnum::DartIntegration::VisualData
+ */
+
+#include <vector>
+
+#include <Corrade/Containers/Optional.h>
+
+#include <Magnum/DartIntegration/DartObject.h>
+#include <Magnum/DartIntegration/visibility.h>
+
+namespace Magnum {namespace Trade{
+    class MeshData3D;
+    class PhongMaterialData;
+    template<UnsignedInt dimensions> class ImageData;
+    typedef ImageData<2> ImageData2D;
+    class TextureData;
+}}
+
+namespace Magnum { namespace DartIntegration {
+/**
+@brief VisualData struct
+*/
+struct VisualData {
+    /** mesh information */
+    Trade::MeshData3D* _meshData;
+    /** material information */
+    Trade::PhongMaterialData* _matData;
+    /** ImageData2D (vector) */
+    std::vector<Trade::ImageData<2>*> _imgData;
+    /** TextureData (vector) */
+    std::vector<Trade::TextureData*> _texData;
```

Comment:
> All types used here are movable (though not copyable) and, based on our previous discussion about uniqueness of the data, I think the structure could just contain everything directly as a value, without a pointer, even the vector -- nothing's ever copied, it's just moved around.
> 
> Feel free to correct me if I'm wrong :)

The only issue that I see is the following (you can help me by-pass it):

```cpp
for (UnsignedInt i = 0; i < importer->textureCount(); ++i) {
    Containers::Optional<Trade::TextureData> textureData = importer->texture(i);
    if (!textureData || textureData->type() != Trade::TextureData::Type::Texture2D) {
        Warning{} << "Cannot load texture, skipping";
        /* we need this to keep the indexing correct */
        data._imgData.emplace_back(nullptr);
        data._texData.emplace_back(nullptr);
        continue;
    }

    Containers::Optional<Trade::ImageData2D> imageData = importer->image2D(textureData->image());
    if (!imageData) {
        Warning{} << "Cannot load texture image, skipping";
        /* we need this to keep the indexing correct */
        data._imgData.emplace_back(nullptr);
        data._texData.emplace_back(nullptr);
        continue;
    }

    data._imgData.emplace_back(new Trade::ImageData2D{std::move(*imageData)});
    data._texData.emplace_back(new Trade::TextureData{std::move(*textureData)});
}
```

In this part of the code, I am reading the textures from the Mesh (`aiScene`). Just before I read the materials. If a material has textures, you can find them using an index. This index needs to be reserved (i.e., even if a texture could not be loaded, we would like all the others to be properly found and indexed). For this reason, I add this `nullptr` in cases where some texture could not be loaded so that the user can make the check and use all the other textures properly.

Another way of doing this, would be to directly alter the indexes of the materials (when needed). Do you have anything else to suggest?

Reply:
Right, sorry, missed this case for some reason.

What about having the structure like this, then?

```cpp
struct VisualData {
    /** mesh information */
    Trade::MeshData3D _meshData;

    /** material information */
    Trade::PhongMaterialData _matData;

    /** ImageData2D (vector) */
    std::vector<Containers::Optional<Trade::ImageData2D>> _imgData;

    /** TextureData (vector) */
    std::vector<Containers::Optional<Trade::TextureData>> _texData;
};
```

That would preserve the "nullability" of particular elements, but also avoids those heap allocations. I thought about reindexing as well, but came to a conclusion that it's better to preserve the "gaps" to indicate that not everything was correctly imported.